p4-16/psa/PSA.mdk

```diff
@@ -469,12 +510,156 @@ have an instance for each entry in the table.
+    }
+}
+
+W next_counter_value(W cur_value, CounterType_t type) {
+    if (type == CounterType_t.packets) {
+        return (cur_value + 1);
+    }
+    // Exactly which packet bytes are included in packet_len is
+    // implementation-specific.
+    PacketLength_t packet_len = <packet length in bytes>;
+    if (type == CounterType_t.bytes) {
+        return (cur_value + packet_len);
+    }
+    // type must be CounterType_t.packets_and_bytes
+    // In type W, the least significant bits contain the byte
+    // count, and most significant bits contain the packet count.
+    // This is merely one example storage format.  Implementations
+    // are free to store packets_and_byte state in other ways, as
+    // long as the control plane API returns the correct separate
+    // packet and byte count values.
+    W next_packet_count = ((cur_value + this.shifted_packet_count) &
+                           this.packet_count_mask);
+    W next_byte_count = (cur_value + packet_len) & this.byte_count_mask;
+    return (next_packet_count | next_byte_count);
+}
+
+void count(in S index) {
+    if (index < this.num_counters) {
+        this.counter_vals[index] = next_counter_value(this.counter_vals[index],
+                                                      this.type);
+    } else {
+        // No counter_vals updated if index is out of range.
+        // See below for optional debug information to record.
+    }
+}
+```
+
+Optional debugging information that may be kept if an `index` value is
+out of range includes:
+
+- Number of times this occurs.
+- A FIFO of the first N out-of-range index valus that occur, where N
+  is implementation-defined (e.g. it might only be 1).  Extra
+  information to identify which `count()` method call in the P4
+  program had the out-of-range `index` value is also recommended.
+
+
 ### Direct Counter
 
 ```
 [INCLUDE=psa.p4:DirectCounter_extern]
 ```
 
+A `DirectCounter` instance must appear in the list of values of the
+`psa_direct_counters` table attribute for exactly one table.  We call
+this table the `DirectCounter` instance's "owner".  It is illegal to
+call the `count` method for a `DirectCounter` instance anywhere except
+inside an action of its owner table.
+
+The counter value updated by an invocation of `count` is always the
+one associated with the table entry that matched.
+
+TBD: How to describe which counter value is updated for tables with
+action profiles and direct counters?  Or should this combination even
+be allowed?
+
+An action of an owner table need not have `count` method calls for all
+of the `DirectCounter` instances that the table owns.
+
+The reference implementation for the `DirectCounter` extern is
+essentially the same as that for `Counter`, including the saturating
+behavior.  Since there is no `index` parameter to the `count` method,
+there is no need to check for whether it is in range.
+
+TBD: What to say about `entry_key` type?  Does it need to be made more
```

Comment:
Generally, for the specs I try to look at the document and not at the IDL code, even when it has comments :) and that's precisely my concern with the control plane API process. 

Looking at the file I can see that the idea is to use the match fields of the tables as the key, although I do not see this being described, e.g. in the comments to the ReadRequest or WriteRequest. This is workable, although there are a number of cases, where this is quite limiting. In many cases, this is not very efficient either, especially when the key is wide. There are advantages too, no doubt about that, for example control plane can be more-or-less stateless if it wishes to (as opposed to the situation where an opaque, fiex-length ID is returned for every entry added to the table). How we are going to converge and how acceptable this will be to a wide range of potential users is another challenge.

Nevertheless, this answers our question about the width of the entry_key -- it's whatever is specified in FieldMatch of the corresponding table plus priority for some tables (again, something not articulated in the code). 

This, and other "trivial" things like the lack of any error code in the WriteResponse or ReadResponse  are precisely the reasons we either need to work with the API group really closely or leave API out for now (but I'd prefer the former option).